### DISCUSSION TITLE
changed-files-ignore-pattern option

### DISCUSSION THREAD

**Discussion by ibezkrovnyi** [pnpm/pnpm#3794] [Ideas]
**pnpm** already has option `test-pattern`, which already does filtration of packages with globs if only files matching to specified test-patterns were changed.

It would be great if **pnpm** had more generic option similar to **Microsoft  Lage**'s `ignore` option **to filter out changed projects, where only specified files were changed**.

```
  /** Which files to ignore when calculating scopes with --since */
  ignore: string[];
```
https://github.com/microsoft/lage/blob/6d6c6be3180ab10938b08a941fb4ed0e9ffa0394/src/types/ConfigOptions.ts#L30-L31

Use cases:
- ignore changes in root-level `.gitlab-ci.yaml`: `ignore[]=.gitlab-ci.yaml` (need to pass [dot](https://github.com/micromatch/micromatch#:~:text=dot) option to micromatch)
- ignore changes in `CODEOWNERS` in any folder: `ignore[]=**/CODEOWNERS` (or `ignore[]=CODEOWNERS`, but first looks more explicit)
- ignore changes in `.github` folder: `ignore[]=.github` (or `ignore[]=.github/**`, again - dot option is required)
- ignore changes in any **markdown** file: `ignore[]=**/*.md` (or `ignore[]=*.md`, but then there will be no way to specify only root-level files)
- ignore some technical folders, reports/logs/cache stored in git repository: `ignore[]={**/scripts,**/reports,**/cache}`
- ignore test stubs: `ignore[]=test/react-*/**`
- ignore test files: `ignore[]=**/tests/**`
- ignore tooling files: `ignore[]=tooling/pipeline-scripts`

This can be option `ignore[]` which expects array of globs, or it can be some advanced syntax for `--filter` (**pnpm** already is able to filter/exclude packages by folders, but currently it expects **folder**===**package** I think)

I think, it should be relative to the workspace root directory, so glob `*.md` matches root-level md files, glob `tooling/pipeline-scripts` matches root-level path `/tooling/pipeline-scripts` and `**/tests` matches directory `tests` in any folder including root-level.

**Comment by zkochan** [comment]
Sounds good but the setting should be probably more explicit. `ignore` is too short and vague

**Reply by ibezkrovnyi** [reply]
- `ignore-files`
- `ignore-changed-files`
- `changed-files-to-ignore`
?

**Comment by ibezkrovnyi** [comment]
@zkochan

1) one more name for option:
- `changed-files-ignore-patterns` (`changedFilesIgnorePatterns`) : `string[]`

2) It looks like main change in logic will be single line:
https://github.com/pnpm/pnpm/blob/632a9e2db28ae4a8a34f24def51f872b7c79d421/packages/filter-workspace-packages/src/getChangedPackages.ts#L66

It should be replaced with:
```ts
  const changedFiles = micromatch.not(diff.split("\n"), changedFilesIgnorePatterns, {
    dot: true,
  });
```

3) Will you accept PR?

Could you please let me know what do you think about the above?

**Reply by zkochan** [reply]
Probably `changed-files-ignore-pattern` to make it similar to `test-pattern`

I am ok with this change. You will probably also need to add the new setting to `@pnpm/config`.
